File: web/src/TokenProvider.ts

export interface ITokenProvider {
    getDefault(): Promise<string>
    get(name: string): Promise<string>
    save(authToken: string, name: string): Promise<void>
    saveDefault(authToken: string): Promise<void>
}

export class MemoryTokenProvider implements ITokenProvider {
    tokens: {[name: string]: string} = {};

    private constructor() {
    }

    private static _instance = new MemoryTokenProvider();
    public static DefaultInstance(): ITokenProvider {
        return MemoryTokenProvider._instance;
    }

    async get(name: string): Promise<string> {
        return Promise.resolve(this.tokens[name]);
    }

    async getDefault(): Promise<string> {
        return Promise.resolve(await this.get("Trinsic"));
    }

    async save(authToken: string, name: string): Promise<void> {
        this.tokens[name] = authToken;
        return Promise.resolve();
    }

    async saveDefault(authToken: string): Promise<void> {
        return this.save(authToken, "Trinsic");
    }
}

File: go/services/token_provider.go
package services

type TokenProvider interface {
	GetDefault() string
	Get(name string) (string, error)
	Save(authToken string, name string)
	SaveDefault(authToken string)
}

type memoryTokenProvider struct {
	tokens map[string]string
}

func (m *memoryTokenProvider) SaveDefault(authToken string) {
	m.tokens["Trinsic"] = authToken
}

func (m *memoryTokenProvider) GetDefault() string {
	return m.tokens["Trinsic"]
}

func (m *memoryTokenProvider) Get(name string) (string, error) {
	if val, ok := m.tokens[name]; ok {
		return val, nil
	}
	return "", nil
}

func (m *memoryTokenProvider) Save(authToken string, name string) {
	m.tokens[name] = authToken
}

func NewMemoryTokenProvider() TokenProvider {
	return &memoryTokenProvider{tokens: make(map[string]string)}
}

var DefaultTokenProvider = NewMemoryTokenProvider()
